### ISSUE TITLE
[v3] errors with setting header row through addSheet/Worksheet and setHeaderRow

### ISSUE THREAD

**Issue by the-vampiire** [theoephraim/node-google-spreadsheet#300]
i came across two bugs when working with the new methods. they are distinct errors but debugging one led to the other so both are described here.

# Issues
- `addSheet` breaks when provided `props.headerValues` as described in `google-spreadsheet` docs
- ~~`setHeaderRow` break due to a range syntax issue~~ see comments below

> related thought: should the `GoogleSpreadsheetWorksheet` [basic properties in the v3 docs](https://theoephraim.github.io/node-google-spreadsheet/#/classes/google-spreadsheet-worksheet?id=basic-sheet-properties) include `headerValues`?
- internally the class doesnt initialize this property unless used (directly or indirectly)
- maybe it should be initialized as null and have a proper getter (and the throw error setter like other props).

## addSheet
- [the v3 docs says](https://theoephraim.github.io/node-google-spreadsheet/#/classes/google-spreadsheet?id=fn-addsheet) to use `props.headerValues`
- only `props.headers` are omitted in `addSheet` which causes the following error

> applies to both `addSheet` and `addWorksheet` as the latter is just a wrapper

```js
async addWorksheet(properties) { return this.addSheet(properties); } // alias to mimic old version
```

> error

```
Google API error - [400] Invalid JSON payload received. Unknown name "headerValues" at 'requests[0].add_sheet.properties': Cannot find field.
```

> cause

```js
async addSheet(properties = {}) {
  const response = await this._makeSingleUpdateRequest('addSheet', {
    properties: _.omit(properties, 'headers'), // <--- fails to omit 'headerValues'
  });
...
  // allow it to work with `.headers` but `.headerValues` is the real prop
  if (properties.headers) { await newSheet.setHeaderRow(properties.headers); }
  if (properties.headerValues) { await newSheet.setHeaderRow(properties.headerValues); }
...
```

> proposed solution

```js
async addSheet(properties = {}) {
...
  // separate header[Value]s as they are not valid update props
  const { headers, headerValues, ...onlyUpdateProps } = properties;
  const response = await this._makeSingleUpdateRequest("addSheet", {
    properties: onlyUpdateProps,
  });
...
  // allow it to work with `.headers` but `.headerValues` is the real prop
  if (headers || headerValues) {
    await newSheet.setHeaderRow(headers || headerValues);
  }
...
```

**Comment by the-vampiire** [comment]
i had to do some digging to figure out the `setHeaderRow` bug. because all the tests in the lib pass but my own usage didnt.

it turns out that if the worksheet `title` has a colon `:` in it **it causes the error**

in retrospect it makes sense because the range is defined by a string:

```js
range: `${this.a1SheetName}!A1`
```

so if the title (sheet name) has a `:` in it it can become

```js
range: `$sheet: name with colon!A1`
```

and somehow breaks the range. no idea why `1:1` fixed it...

**Comment by the-vampiire** [comment]
this is obviously an edge case not even google considered (cant find any warning in their docs).

i tried with `/` and `-` and both worked

- do you think its worth updating the docs to include a warning about the title?
- or should we add a regex test to check the title and throw an error? it would be near impossible to figure this out otherwise

**Comment by theoephraim** [comment]
Thanks for doing this debugging and digging :)
I think there is likely a way to encode any colon in the title that would make it work as intended.
I'll take a look at this tomorrow and see if I can get it all working.

Thanks

**Comment by theoephraim** [comment]
fixed the header/headerValues stuff. It should really be a getter like the other props, but not a high priority...

Regarding the colon issue, I thought I'd be able to figure out a way to encode it, and I did a lot of digging. Turns out it is a bug on google's side that only affects some api calls and not others. I left an issue here: https://issuetracker.google.com/issues/150373119

I also added an error in the module with a nice message to help.

My advice - avoid colons in sheet titles!
